### ISSUE TITLE
Support ES6 arrow syntax in psc output for currying

### ISSUE THREAD

**Issue by wegry** [purescript/purescript#2204]
In Purescript 0.8, a Purescript function like `w x y z = z` is output as 
```
var w = function (x) {
     return function (y) {
        return function (z) {
            return z;
        };
    };
};
// 122 bytes
```
An option would be added to psc to output this instead with `=>` and instead be output as the following: `var w = x => y => z => z; // 25 bytes`

It might be easier to have the final function always be wrapped in curly braces:
```
var w = x => y => z => { 
    return z;
}; // 42 bytes
```

An identity function would go from 
```
var id = function (x) {
    return x;
}; // 40 bytes
```
to `var id = x => x; // 16 bytes`

**Comment by garyb** [comment]
I think that's a pretty micro optimisation with the downside of making the JS incompatible with a whole bunch of browsers, without using something like babel which will re-bloat the JS :smile:. I guess we could provide a flag to choose a codegen version or something.

gzip makes a pretty large difference too. We have an app that generates a 3.4mb JS file, that reduces to 449kb (13% of its original size) when served with gzip enabled.

**Comment by wegry** [comment]
@garyb Not if you're only supporting the [release Chrome](https://kangax.github.io/compat-table/es6/) :wink:. My team currently generates a bundle of ~7mb worth of JS, 45% of which, and that share is growing, is compiled Purescript. Removing our the use of `psc-bundle` alongside webpack@2.1 and uglify gets us down to ~600kb gzipped (as opposed to 1.1mb with psc-bundle inside webpack).

**Comment by garyb** [comment]
That's a fair example! I guess my point was I don't think we should enable ES6 output by default if we add it.

The generated code should look pretty nice with it enabled though...

**Comment by kRITZCREEK** [comment]
Without `language-javascript` supporting ES6 it seems a bit silly to not allow arrow syntax in FFI modules, but generating it from the compiler...

**Comment by wegry** [comment]
I agree on it not being the default. I do think the relative ease of something like this, as opposed to something like ES6 module support which I'm hoping is still coming at some point, makes it worth taking a shot at.

**Comment by spicydonuts** [comment]
I'm actually in the ES6 camp.  Support is getting much better so it isn't that unusual to not need to transpile, and at the same time ES6 => ES5 compilers are pretty much the norm now.

I don't think that's the right way to look at it though.. It seems better to target the general JS ecosystem and tooling rather than a specific runtime, whether it be legacy browsers or Node with CommonJS.  Modern JS tooling works best with ES6 modules.  They allow Webpack and Rollup (and more tools in the future) to perform static analysis and remove unused code.  It's also the format future JS VMs will use, so tooling can be removed over time as it's no longer needed and the code stays the same.

Here are the pros I see with full ES6 output from `psc`:
1. simplify or completely remove `psc-bundle`
2. remove custom DCE code which only works for some use cases (doesn't work with purs-loader, for example) in favor of existing tools dedicated to this job
3. cleaner integration with other languages and tooling -- TypeScript uses modules, CLJS can import them, etc
4. make use of PS' module and type information 

I also don't think this will complicate the current default project setup.  Pulp already uses Webpack and Browserify (at least last time I looked at it).  It could just use Webpack for both the dev server _and_ bundling/optimizing, with a basic Babel config for generating the same browser support it has now (or better actually, seeing how someone could upload a PS module right now with with `const` in its FFI).

**Comment by wegry** [comment]
@spicydonuts I do agree psc-bundle is not nearly as robust as Webpack. My team is hoping to remove psc-bundle from our build in the near future for a fairly massive reduction in bundle size. 

@kRITZCREEK the FFI files should probably not try to parse for a specific language level and instead _only_ look for either CommonJS or ES6 exports. It would allow teams to use ES6 imports (for tree shaking in the eventual bundling process). My team currently deals with the FFI's limitations by writing tiny modules, a la lodash, for PS that sits on top of JS. JS that depends on PS can utilize tree shaking. 

purs-loader allows more robust JS tooling, like eslint, to work with FFI files which we also hope to switch to. It might be something that would comprise the normal workflow for PS projects.

**Comment by hdgarrood** [comment]
@spicydonuts @wegry Can you give examples where psc-bundle performs worse than other (JS) dead code elimination tools on compiled purescript code? All experiments I've seen had the opposite result.

**Comment by hdgarrood** [comment]
Also, I don't understand why you need to be using purs-loader to use eslint on FFI files, could you explain where you're coming from there please?

**Comment by garyb** [comment]
> Also, I don't understand why you need to be using purs-loader to use eslint on FFI files, could you explain where you're coming from there please?

I also wonder what the problem is here - we use jscs and jshint on the FFI files for the core libraries as part of the build.

**Comment by hdgarrood** [comment]
Regarding Pulp, Pulp only uses Webpack in `pulp server` (and it doesn't really need to, actually I want to drop the Webpack dependency entirely, see bodil/pulp#151), and it only uses Browserify in `pulp browserify`, which is just one solution amongst many for bundling up purescript code together with javascript code from npm into a browser bundle. A significant chunk &mdash; perhaps even the majority &mdash; of Pulp usage doesn't use Webpack or Browserify at all. I'm not sure I completely understand your proposal, but I think I understand it well enough to say it would significantly complicate how Pulp works, so I'm not really very keen.

By the way, are you using both Pulp and purs-loader, and if so, why?

**Comment by wegry** [comment]
@garyb Because it's what we use for the rest of our JS.

**Comment by natefaubion** [comment]
> 2. remove custom DCE code which only works for some use cases (doesn't work with purs-loader, for example) in favor of existing tools dedicated to this job

`purs-loader` has worked with psc-bundle for a while now.

**Comment by spicydonuts** [comment]
This was my main point, and I feel like it was misunderstood since it's my answer to everything above:

> it seems better to target the general JS ecosystem and tooling rather than a specific runtime, whether it be legacy browsers or Node with CommonJS

PS would be more focused and more pluggable into other tooling if it targeted ES6 and let existing tools take care of the rest, including browser support/platform targeting, bundling, minifying, code splitting, and tree shaking.

Below are more specific replies as well.  I don't mind going into detail on them as long as it's to flush out concerns or clarify, but none I don't want to get caught in the weeds down there and miss talking about the general idea.

> `purs-loader` has worked with psc-bundle for a while now.

Do you mean that if I have `A.js` and `B.purs` in a folder and `A.js` contains `import {foo} from './B'`, and `B.purs` contains definitions for both `foo` and `bar`, that the compiler will leave `bar` out of the compiled JS?  That's what we'd get using ES6 modules, and it works for every language that compiles to ES6 modules.

> @spicydonuts @wegry Can you give examples where psc-bundle performs worse than other (JS) dead code elimination tools on compiled purescript code? All experiments I've seen had the opposite result.

If you mean an example of PS compiling to ES6 modules so we can compare Webpack's tree shaking to PS' DCE, then no.  If you mean an example of some JS project using ES6 modules to remove dead code then yes, but I'm not sure how that would help much when comparing to PS' DCE.  PS' DCE doesn't do anything for the JS required by browserify.

> Also, I don't understand why you need to be using purs-loader to use eslint on FFI files, could you explain where you're coming from there please?

> I also wondered about that - we use jscs and jshint on the FFI files for the core libraries as part of the build.

No one _needs_ to use Webpack to hook up their linter, but it's a very popular way to do it.  The idea is instead of having a separate linting configuration with its own targets and file watchers you just tell Webpack that each file it's about to process needs to be run through the linter first.  If PS had a linting tool for formatting or something you could plug that in as well and it would all run through the same pipeline.

Again, the point isn't that everyone should do it that way; the point is that we would support more use cases with less code by targeting the ecosystem instead of having PS tools run the whole compile/bundle/optimize process.

> I'm not sure I completely understand your proposal, but I think I understand it well enough to say it would significantly complicate how Pulp works, so I'm not really very keen.

What are you concerned about?  The parts that currently don't use Webpack or Browserify would continue to not use Webpack or Browserify.  The parts that do could be simplified to use the same dependencies and configuration, by using Webpack for everything or Browserify for everything (though the Browserify route will mean more custom scripts and potentially something like Gulp and Rollup.. or it could just use Webpack.. just an example though).

> By the way, are you using both Pulp and purs-loader, and if so, why?

No, it was just an example of the change benefiting Webpack, Rollup, and other JS tools while also pointing out that it wouldn't change the current Pulp experience.

**Comment by garyb** [comment]
Fair points, I pretty much only work in PS projects where the only JS is FFI code. And I don't interact with a watch-based build process as SlamData became to large for that quite a while ago.

Two points that are not meant as arguments, just considerations worth mentioning:
- I think any non-PS-specific DCE method will suffer a little in a comparison with `psc-bundle`, given that it has special knowledge of PS-generated-JS. I seem to remember some tool claiming our generated code for data constructors is impure so can't be erased.
- `psc-bundle` is where the uncurrying will be enabled once that PR is merged (it operates on the output as a whole, not module-at-a-time), so I'm not sure how that could be integrated into a webpack build situation

Maybe there's a solution for both of the above that would allow those things to play well with Webpack, Rollup, etc. but I don't know enough about their architectures and capabilities to say.

**Comment by spicydonuts** [comment]
Ah, interesting.  I wonder if there's a way to fix that.. ideally the dev-mode rebundle would be all in memory and only rebuild the piece that changed (and dependencies).

> I think any non-PS-specific DCE method will suffer a little in a comparison with `psc-bundle`

This is definitely true for CommonJS module optimizers.. they can't do very much safely.  What information besides module imports and exports does `psc-bundle` use?  It'd be good to outline exactly what information would be lost if a PS module is compiled to an ES6 module.  I can't think of anything but I might not know all the details..

> `psc-bundle` is where the uncurrying will be enabled once that PR is merged (it operates on the output as a whole, not module-at-a-time)

Does it use any information to do that that isn't available in the JS AST?  If not, that could still be a useful optimization to apply post-build to any JS bundle, regardless of whether it came from PS or not (but PS-compiled JS would benefit the most).  I have no idea how it works so that could be way off.  If that's true though, it sounds like a great generic JS tool.

**Comment by garyb** [comment]
>  I wonder if there's a way to fix those build times..

Possibly, but it's not something I have much of an issue with - I enjoy near-instant rebuilds from `psc-ide` for the file I'm working on, and then trigger a manual build when I actually have something I need to check in the browser.

We should be able to have much quicker rebuilds soon thanks to scanning module headers when rebuilding rather than having to re-parse everything. It takes 35 seconds to rebuild with no changes currently!

The bundling processes take a while too, a few seconds each for psc-bundle and webpack's bundling. Maybe that would be faster with a watch - I think the webpack bundling is incremental then? Webpack is already way faster than browserify was though.

I don't really expect lightning fast rebuilds with a 920-module project though :laughing: 

> This is definitely true for CommonJS module optimizers.. they can't do very much safely. What information besides module imports and exports does `psc-bundle` use?

Having glanced at the code again, `psc-bundle` isn't actually doing anything particularly clever. I think there's only one assumption that does not apply to arbitrary JS: every definition is pure. Here's the issue I had in mind: #2177.

> Does it use any information to do that that isn't available in the JS AST?

No, anything in `psc-bundle` has no information other than the JS it's provided with, but maybe there are assumptions like the prior-mentioned purity in how it operates - I'm not sure really, I've not looked too closely at the details of that PR.

**Comment by wegry** [comment]
>  I think any non-PS-specific DCE method will suffer a little in a comparison with psc-bundle, given that it has special knowledge of PS-generated-JS. I seem to remember some tool claiming our generated code for data constructors is impure so can't be erased.

Maybe you're referring to this? @garyb purescript/purescript/issues/2177

Uglify (without psc-bundle) on my large project has about 1/4 of the errors pulling in PS CommonJS modules relative to pulling in psc-bundle as a whole.

**Comment by garyb** [comment]
^ That's the one, I just mentioned so above :smile:

**Comment by wegry** [comment]
Whoops, saw the next response too late 😅

**Comment by paf31** [comment]
I'll weigh in on this soon, but I just want to say that it's possible to get a fast browser development workflow (on par with `psc-ide`) if you work on one module at a time. This is what Try PureScript and the new PSCi do: bundle all dependencies once using `psc-bundle`, and reload the one you're working on using some `require` shim.

**Comment by AppShipIt** [comment]
> I seem to remember some tool claiming our generated code for data constructors is impure so can't be erased.

Cf. https://github.com/purescript/purescript/issues/1206#issuecomment-195143286 via @natefaubion. Discussing with @spicydonuts in the office. If we could make data constructors work with Rollup and/or WebPack 2 would you be interested in incorporating their shape into #2205?

**Comment by garyb** [comment]
It's kind of a separate issue, #2205 is about unifying the representation of type classes and data constructors in `CoreFn` onwards. But sure, if we can come up with and IIFE-free equivalent to the constructor definitions I'd have no problem with opening a PR for that.

I'll open a new issue to discuss.

**Comment by spicydonuts** [comment]
If `psc-bundle` doesn't use anything except static import/export for DCE then a few small tweaks to the data type JS would fix that.  There's even an ES6 syntax for that. 😉

An ES6 `data` type might look like this, and should be statically tree-shakeable:

```js
export class Just {
  constructor (private value0) { }
  static create (value0) {
    return new Just(value0);
  }
}
```

That just leaves the curry-optimizing.. If it's already a generic JS tool we should definitely rebrand it and make it easy for any curry-heavy languages to benefit from it.

Side thought.. it _looks_ like the constructor pattern used in the current data IIFEs tries to make "new" work normally in JS (non-curried).  We _could_ continue that theme with an extra field to allow JS to see if an instance is a Maybe:

```js
export const Maybe = {}

export class Just {
  constructor (private value0) { }
  static create (value0) {
    return new Just(value0);
  }
  $type = Maybe
}
```

Not super related though 🙂

**Comment by spicydonuts** [comment]
Or inheritance:

```js
export class Maybe { }

export class Just extends Maybe {
  constructor (private value0) {
    super()
  }
  static create (value0) {
    return new Just(value0);
  }
}
```

But that might encourage a JS person to add their own implementation which doesn't make sense.

**Comment by spicydonuts** [comment]
```js
import {Maybe, Just} from 'Data.Maybe'

const x = new Just(9)

console.log(x instanceof Maybe)
console.log(x instanceof Just)
```

..ok, I'll stop now 😆

**Comment by garyb** [comment]
Yeah, I was going to suggest that also. I'm not sure we should allow too much JS runtime reflection on the PS representations though... although they've not changed in a really long time, I guess they might do so in the future. Using the same functions PS does to interact with them is preferred :wink:

**Comment by spicydonuts** [comment]
I definitely understand the desire to keep FFI opaque enough that people realize they shouldn't rely on it, but if a concept is nearly one-to-one with a JS idiom (the simplest example probably being functions) then I don't think it makes sense to make it more difficult.

**Comment by wegry** [comment]
@spicydonuts Is there a private keyword for ES6 constructor parameters?

**Comment by AppShipIt** [comment]
@wegry Only in TypeScript.

**Comment by hdgarrood** [comment]
Thanks @wegry and @spicydonuts, I understand much better where you're coming from now.

I'm not really convinced by the original argument re: size, because I think the answer there ought to be configuring your server to use compression, in which case I doubt you'd see any significant difference between the two styles - I'd expect the repeated keyword "function" to be very easily compressable, for instance. @wegry: can you elaborate a little about what you changed in your build process to see that difference? Is the difference perhaps due to you using a different DCE tool, which was able to operate on your entire application, as opposed to `psc-bundle`, (which, as you've noted, isn't fantastic in some cases where you have lots of both PS and JS)?

I agree with the majority of what people are saying here, though, and it seems like the conversation has shifted and we're really talking about using ES6 modules for imports and exports now. Is that correct? If so, can we close as a duplicate of #1206?

**Comment by AppShipIt** [comment]
I'd contend arrows is subsumed by #2207 too 👍

**Comment by wegry** [comment]
@hdgarrood Compression for serving up files is something we are already doing and pretty common these days. It's also not a substitute for cutting waste from psc output. It can only alleviate the size penalty of using Purescript.

We cannot use DCE with psc-bundle, because our PS depends on JS and vice-versa. Uglify minification with webpack@2.1 bundling psc generated CommonJS modules and ES6 is what allows for the best output size for us.

What we had before:

1. psc generates CommonJS modules
2. psc-bundle wraps these all into an unminified bundle
3. We have a node script wrap psc-bundle's bundle in an AMD module // obviously not ideal, but this was an AMD project prior to my webpack work.
4. Webpack picks up the AMD module as a requirement for about ~20ish JS files

_1.1MB output minified (with uglify) and gzipped_

After removing the psc-bundle:

1. psc generates Common.js
2. Webpack directly picks psc CommonJS module output for ES6 (and AMD modules depending on them).

_550KB output minified (with uglify) and gzipped_

The only 'dead' code I've personally seen removed is psc-bundle mangling external requires from the FFI, but we're probably not the intended use case.

```
// How we currently import ES6 modules into the FFI.
// Wrapped in a closure
exports.k = function (x) {
    return require('../isModule').default(x);
}; 
```

```
// Doesn't work with psc-bundle but does without
var isModule = require('../isModule').default

exports.k = function (x) {
    return isModule(x);
};
```

I believe rollup uses uglify for minification as well and is probably the best thing to target. @garyb seems to suggest that there isn't much in the way of PS specific DCE going on, and honestly, with CommonJS modules, I don't think that's possible.  If the team insists on rolling their own solution, webpack or rollup with uglify is a good bar to hold your DCE/minifier too as well.

@AppShipIt I agree on this being subsumed.

**Comment by wegry** [comment]
Also @hdgarrood I apologize this has turned into a duplicate of #1206 and I'll see myself out. 😅

**Comment by paf31** [comment]
DCE is very important for my use case, and I use `psc-bundle` and get great results. I'd like to see an example demonstrating that Webpack/Uglify/whatever can match psc-bundle in the best case. Hello World would be a good test to start with.

Also, I'm keen on the idea of moving to ES6 codegen in general, if we can do it without losing features. My concern is that people use PS in a lot of ways, not all alongside a modern JS toolchain. Disregarding those other workflows would be a mistake, I think.

**Comment by wegry** [comment]
@paf31 I don't doubt its utility and my team has used it for quite some time. But have you compared it to any other tooling? If you'd like to provide the full, minified, and minified + gzipped size of psc + webpack vs psc + psc-bundle with the ideal greenfield example (PS on top of JS) for reference, that'd be great.

I think 'DCE' is a bit opaque too. It seems that a lot of things that `psc` can only do when you've got PS modules still, like remove unused variables, is not done. Or is psc-bundle only loading used modules? If there's documentation about this somewhere, please point me to it. 

Hello World is pretty atrocious as far as space is concerned with Webpack (which has its own overhead from bundling). I'd propose that a fairer benchmark would, with a JS entry point.

##### Use Case 1: webpack@2.1 with psc-bundle
`entry.js` depends on the psc-bundle of `purs-example.purs` depending on `Prelude`. 

##### Use Case 2: webpack@2.1 without psc-bundle
`entry.js` depends on `./output/purs-example` (transpiled) depends on `./output/Prelude` (transpiled).

My team is currently on 0.8, but its my understanding that purescript@^0.9 has broken up the Prelude).

**Comment by paf31** [comment]
> If you'd like to provide the full, minified, and minified + gzipped size of psc + webpack vs psc + psc-bundle with the ideal greenfield example (PS on top of JS) for reference, that'd be great.

I suggest someone who wants to convince me JS tooling can match psc-bundle provide the evidence for that 😄 

The other problem with dropping psc-bundle is that it is currently the most sensible place to put global program optimizations. Admittedly, we don't have any yet, but we will hopefully have one soon (see #2139).

**Comment by AppShipIt** [comment]
> I have high hopes for Rollup in particular.

In theory WebPack 2 is comparable but we're still testing locally.

**Comment by wegry** [comment]
Lame example of psc-bundle generating more code than webpack@2.1 I whipped up in half an hour:

```
//with-psc-bundle.js
import PS from './output/PS' //entry point being Example.purs

console.log(PS['Example'].workIt(5))
```
```
//without-psc-bundle.js
import {workIt} from './output/Example'

console.log(workIt(5))
```

```
--Example.purs
module Example where

import Prelude --only part of this is used (+)
import Data.Void --none of this is used

workIt :: Int -> Int
workIt x = x + 3
```

_Results?_
```
          bundle  minified  gzipped + minified
with      56.3kB    23.8kB               5.0kB   
without   54.3kB    21.4kB               4.7kB
```

Caveats, this understates the fact that purescript needs a way to effectively deal with unused dependencies. It's also too small for the benefits of minifying with uglify to really shine through. My team is currently buried in about 2K warnings from updating from 0.7 to 0.8. Most of these are about specifying imports. Ideally, psc would remove imports that are not used. psc-bundle definitely doesn't currently.

I think the uncurrying optimization will be wonderful @paf31. That said, I think it should be in psc for the same reason as removal of unused variables.

**Comment by paf31** [comment]
The uncurrying optimization can't be in `psc` unfortunately. It won't play nicely with separate compilation.

**Comment by hdgarrood** [comment]
I do quite like the idea of having the uncurrying optimization as a completely separate program that operates on arbitrary ES6 modules, though, if that's doable. We might even write it in purescript &mdash; then, we'd have access to a ton of existing tooling written in JS for stuff like parsing ES6.

**Comment by kevinbarabash** [comment]
I have a small project test project that implements a calculator using React.  I've been experimenting using some code from [rollup-plugin-purs](https://github.com/Pauan/rollup-plugin-purs) to convert the current output of `purs compile` into ES6 modules.  I then use `rollup` to bundle them.  It does an okay job of removing dead code, but there are some things it could remove but doesn't.  I wrote a small script that does a couple of AST transforms on the rollup's output to remove even more dead code.  Here are the tests results (note these are all uncompressed numbers):
```
purs bundle: 2,518 K
rollup after converting to ES6: 74 K
additional processing on rollup bundle: 14 K
```
The additional processing includes:
- removal of dead code (I create a dependency tree from the entry point and filter out all top level declarations that aren't in that tree, this got the bundle down to 25 K)
- removal of comments (not really a big win since these would disappear with uglification)
- transformed all function expressions not used for data constructors to be array expressions
- inlining of `mul(semiringNumber)` to `_numMul` and similar constructors (my app passes `(*)` and other operators as arguments to other functions

source: https://github.com/kevinbarabash/purescript-calculator-example
demo: https://kevinbarabash.github.io/purescript-calculator-example

**Comment by kevinbarabash** [comment]
The numbers I posted above for `purs bundle` are a lot higher than they should be.  I didn't realize you had to specify the `-m, --main` option for it to do dead code removal.  After doing that, the bundle size is 54K.

**Comment by chrisdone** [comment]
Having done a lot of [down and dirty minification with Fay and closure](https://github.com/faylang/fay/wiki/Reducing-output-size), I took a look at a colleague's problem with output size when using Halogen. The app takes 300KB after minification (pre-compression), with either Uglify or Closure, of the result of `purs bundle`. whereas the equivalent app written in Elm is 20-30KB. This indicated there was some dead code removal not working properly.

Picking on a single module `Record.Unsafe`, there's an obvious "low-hanging fruit" that's likely causing the bulk of the output size. The particular output format used by PureScript's codegen at the moment is blocking some pretty trivial code stripping that uglifyjs or closure could normally do:

![Screenshot from 2019-07-28 09-41-39](https://user-images.githubusercontent.com/11019/62004494-ce034700-b11d-11e9-8209-831100ea70a7.png)

It basically just removed whitespace. Now, if we remove those strings for module names and just use variable names, like I did in Fay, well then Uglify deletes all the code, as it should:

![Screenshot from 2019-07-28 09-46-54](https://user-images.githubusercontent.com/11019/62004514-115db580-b11e-11e9-876b-965121d7501f.png)

Similarly, using a single function from the module, Closure can inline that one function:

![Screenshot from 2019-07-28 09-47-37](https://user-images.githubusercontent.com/11019/62004524-22a6c200-b11e-11e9-8163-8e01aec9406e.png)

So I'd propose that the codegen should have two modes; one mode for "separate compilation", which is the default mode you already have, and then another mode for "single compilation", let's say, which would cater to minification tools like UglifyJS2 (the author of which is a good friend) or Google Closure. This would also permit uncurrying of saturated functions, which would reduce that Closure example further.

**Comment by hdgarrood** [comment]
@chrisdone is your colleague using `purs bundle`? I suspect it would be able to remove that code.

**Comment by chrisdone** [comment]
Yes, as I wrote in my comment.

**Comment by hdgarrood** [comment]
Can you please put together a self-contained example including the commands used? I am fairly sure that `purs bundle` should be able to handle this.

**Comment by hdgarrood** [comment]
The output you have there looks like it may have been produced using something like `purs bundle --module Record.Unsafe`, in which case it is intended that you get everything in the module.
